docs/practices/appeng/adrs/0005-javascript-dependency-management.mdx

```diff
@@ -0,0 +1,273 @@
+are always in-play. These files include a package file and a lock file which
+stores the version installed of these packages. This decision covers some of the
+ways to do this in either _Yarn_ or _NPM_.
+
+### Dependency resolution
+
+Regardless of the JDMT being used, there will always be a `package.json` file
+which will contain the three different sections for dependencies. When it comes
+to installing dependencies in these sections, ensure that any commands for the
+JDMT leverage the exact version that you are trying to install.
+
+<Tabs groupID="JDMT">
+<TabItem value="npm" label="npm-cli">
+
+```shell title="installing all dependencies in NPM"
+>_ npm install
+```
+
+</TabItem>
+<TabItem value="yarn" label="classic yarn-cli">
+
+```shell title="installing all dependencies in Yarn"
+>_ yarn
+```
+
+</TabItem>
+</Tabs>
+
+#### Dependencies
+
+- listed in `"dependencies": {}` in `package.json`
+- needed for the site work
+- example: the site won't work in production without React
+
+<Tabs groupID="JDMT">
+<TabItem value="npm" label="npm-cli">
+
+```shell title="installing react version 1.0.1 exactly using NPM as a dependency"
+>_ npm install --save-exact react@1.0.1
+```
+
+</TabItem>
+<TabItem value="yarn" label="classic yarn-cli">
+
+```shell title="installing react version 1.0.1 exactly using Yarn as a dependency"
+>_ yarn add react@1.0.1
+```
+
+</TabItem>
+</Tabs>
+
+#### Development Dependencies
+
+- listed in `"devDependencies": {}` in `package.json`.
+- used in development, but not production
+- example: we use Prettier to format code
+
+<Tabs groupID="JDMT">
+<TabItem value="npm" label="npm-cli">
+
+```shell title="installing prettier version 1.0.1 exactly using NPM as a development dependency"
+>_ npm install --save-exact --save-dev prettier@1.0.1
+```
+
+</TabItem>
+<TabItem value="yarn" label="classic yarn-cli">
+
+```shell title="installing prettier version 1.0.1 exactly using Yarn as a development dependency"
+>_ yarn add --dev prettier@1.0.1
+```
+
+</TabItem>
+</Tabs>
+
+#### Peer Dependencies
+
+- listed in `"peerDependencies": {}` in `package.json`
+- used to express the compatibility of your package with a host tool or library
+    - typically only found in packages intended to be published to _NPM_
+- works with or alongside another library
+- example: `@material-ui/icons` works with `@material-ui/core`
+
+#### Version locking
+
+You will notice that all of the packages listed are set to specific versions
+(instead of a range using the `^` or `~` characters). This is to help ensure
+that installing dependencies is deterministic and it also means we are in full
+control of all version updates.
+
+By default, JDMT will use the `^` caret character for the version which does not
+lead to control of what version get installed when running the JDMT installation
+command. Avoid using semantic version ranges when install dependencies by using
+some of the commands listed above around installation.
```

Comment:
For this section, I think we should have a recommendation for using `.npmrc` file to set the `--save-exact` flag to be true for _NPM_. This way we don't need to rely on folks having the correct flags when running `npm install` commands. I don't believe there is an equivalent _Yarn_ file, but I will look after posting this.

[=> See docs here](https://docs.npmjs.com/cli/v8/using-npm/config)

Reply:
Nothing to configure _Yarn_ to leverage the same kind of behavior as _NPM_ here. But there is a `.yarnrc` file that exists.

[=> _Yarn_ docs on configuration](https://classic.yarnpkg.com/en/docs/yarnrc)